### PR TITLE
adding xd75 keymap

### PR DESCRIPTION
**Pull request by martinmksiu**
QWERTY keymap for xd75 with centered numpad and standard 40% layers for easy switching between xd75 and contra/planck

### CODE REVIEW

File: keyboards/xd75/keymaps/martin_cennum/keymap.c [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* What this layout is: QWERTY layout with the standard 40% layers. For easy
+ * switching between my xd75 and contra
+ */
+
+#include "xd75.h"
```
Comment (mechmerlin) [inline keyboards/xd75/keymaps/martin_cennum/keymap.c]: can you change this line to `#include QMK_KEYBOARD_H`

File: .directory [hunk 2]
```diff
@@ -0,0 +1,4 @@
+[Dolphin]
```
Comment (drashna) [inline .directory]: Could you remove this file?

File: keyboards/contra/keymaps/martin/config.h [hunk 3]
```diff
@@ -0,0 +1,42 @@
+#ifndef CONFIG_USER_H
```
Comment (drashna) [inline keyboards/contra/keymaps/martin/config.h]: Could you replace the first 4 lines, eg, these lines: 
```c
#ifndef CONFIG_USER_H
#define CONFIG_USER_H

#include "config_common.h"
```
with just `#pragma one`, and remove the endif at the end?

File: keyboards/contra/keymaps/martin_bac/config.h [hunk 4]
```diff
@@ -0,0 +1,42 @@
+#ifndef CONFIG_USER_H
```
Comment (drashna) [inline keyboards/contra/keymaps/martin_bac/config.h]: Could you replace the first 4 lines, eg, these lines: 
```c
#ifndef CONFIG_USER_H
#define CONFIG_USER_H

#include "config_common.h"
```
with just `#pragma one`, and remove the endif at the end?

File: keyboards/xd75/keymaps/martin_cennum/config.h [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef CONFIG_USER_H
```
Comment (drashna) [inline keyboards/xd75/keymaps/martin_cennum/config.h]: Could you replace the first 4 lines, eg, these lines: 
```c
#ifndef CONFIG_USER_H
#define CONFIG_USER_H

#include "config_common.h"
```
with just `#pragma one`, and remove the endif at the end?

File: keyboards/xd75/keymaps/martin_cennum/rules.mk [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+BACKLIGHT_ENABLE = yes       # Enable keyboard backlight functionality
+
+ifndef QUANTUM_DIR
```
Comment (drashna) [inline keyboards/xd75/keymaps/martin_cennum/rules.mk]: Could you remove this block?  It's not needed and a holdover from the old makefile system.

File: keyboards/contra/keymaps/martin_bac/keymap.c [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "contra.h"
```
Comment (drashna) [inline keyboards/contra/keymaps/martin_bac/keymap.c]: Could you replace this with just `#include QMK_KEYBOARD_H`

File: keyboards/contra/keymaps/martin_bac/config.h [hunk 8]
```diff
@@ -0,0 +1,37 @@
+#pragma one
```
Comment (drashna) [inline keyboards/contra/keymaps/martin_bac/config.h]: This should be `#pragma once`, not "one".  (that may be my fault, but)

File: keyboards/contra/keymaps/martin_bac/keymap.c [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+   * | Tab  |   "  |   ,  |   .  |   P  |   Y  |   F  |   G  |   C  |   R  |   L  | Bksp |
+   * |------+------+------+------+------+-------------+------+------+------+------+------|
+   * | Esc  |   A  |   O  |   E  |   U  |   I  |   D  |   H  |   T  |   N  |   S  |  /   |
+   * |------+------+------+------+------+------|------+------+------+------+------+------|
+   * | Shift|   ;  |   Q  |   J  |   K  |   X  |   B  |   M  |   W  |   V  |   Z  |Enter |
+   * |------+------+------+------+------+------+------+------+------+------+------+------|
+   * | Ctrl |Hyper | Alt  | GUI  |Lower | Func |Space |Raise | Left | Down |  Up  |Right |
+   * `-----------------------------------------------------------------------------------'
+   */
+  [_DV] = {
```
Comment (drashna) [inline keyboards/contra/keymaps/martin_bac/keymap.c]: This should be converted to use `LAYOUT_ortho_4x12` here

File: keyboards/xd75/keymaps/martin_cennum/config.h [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma one
```
Comment (drashna) [inline keyboards/xd75/keymaps/martin_cennum/config.h]: This should be `pragma once` specifically

File: keyboards/xd75/keymaps/martin_cennum/keymap.c [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+ * |--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------|
+ * | ESC    | A      | S      | D      | F      | G      | 4      | 5      | 6      | H      | J      | K      | L      | ;      | '      |
+ * |--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------|
+ * | LSHIFT | Z      | X      | C      | V      | B      | 1      | 2      | 3      | N      | M      | ,      | .      | /      | ENTER  |
+ * |--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+-----------------+--------+--------|
+ * | LCTRL  | HYPER  | LALT   | LGUI   | LOWER  | FN     | 0      |        | .      | SPACE  | RAISE  | LEFT   | DOWN   | UP     | RIGHT  |
+ * '--------------------------------------------------------------------------------------------------------------------------------------'
+ */
+
+ [_QW] = { /* QWERTY */
```
Comment (drashna) [inline keyboards/xd75/keymaps/martin_cennum/keymap.c]: This should be converted to `LAYOUT_ortho_5x15`

File: keyboards/xd75/keymaps/martin_cennum/rules.mk [hunk 12]
```diff
@@ -0,0 +1,5 @@
+
+
+ifndef QUANTUM_DIR
```
Comment (drashna) [inline keyboards/xd75/keymaps/martin_cennum/rules.mk]: This block should be removed, as it's a hold over from the old makefile system

File: keyboards/contra/keymaps/martin_bac/config.h (hunk 8 above)
Comment (drashna) [reply keyboards/contra/keymaps/martin_bac/config.h]:
```suggestion
#pragma once
```

File: keyboards/xd75/keymaps/martin_cennum/config.h (hunk 10 above)
Comment (drashna) [reply keyboards/xd75/keymaps/martin_cennum/config.h]:
```suggestion
#pragma once
```

**Comment by drashna** [comment]
Just wanted to touch base here.  Have you had a chance to update your PR?

File: keyboards/contra/keymaps/martin/keymap.c [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include QMK_KEYBOARD_H
+
+#define _______ KC_TRNS
```
Comment (drashna) [inline keyboards/contra/keymaps/martin/keymap.c]: This line isn't needed, as it's part of quantum_keycodes.h. 

You can (and should) remove this.

```suggestion
```

File: keyboards/contra/keymaps/martin_bac/keymap.c [hunk 14]
```diff
@@ -0,0 +1,107 @@
+#include QMK_KEYBOARD_H
+
+
+// Fillers to make layering more clear
+#define _______ KC_TRNS
```
Comment (drashna) [inline keyboards/contra/keymaps/martin_bac/keymap.c]:
```suggestion
```

This should be removed.

File: keyboards/xd75/keymaps/martin_cennum/keymap.c [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+ */
+
+/* What this layout is: QWERTY layout with the standard 40% layers. For easy
+ * switching between my xd75 and contra
+ */
+
+#include QMK_KEYBOARD_H
+
+// Fillers to make layering more clear
+#define _______ KC_TRNS
```
Comment (drashna) [inline keyboards/xd75/keymaps/martin_cennum/keymap.c]:
```suggestion
```

This should be removed.

File: keyboards/xd75/keymaps/martin_cennum/keymap.c [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+  { _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______  },
+  { RESET  , _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______  },
+ },
+};
+
+const uint16_t PROGMEM fn_actions[] = {
+
+};
+
+const macro_t *action_get_macro(keyrecord_t *record, uint8_t id, uint8_t opt)
```
Comment (drashna) [inline keyboards/xd75/keymaps/martin_cennum/keymap.c]: Would you mind converting this to `process_record_user`? 

You should see an example of this in the planck default keymap.

Comment (noroadsleft) [reply keyboards/xd75/keymaps/martin_cennum/keymap.c]: This code block can be safely deleted; the keymap isn't using it.

File: keyboards/xd75/keymaps/martin_cennum/keymap.c [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+ [_FN] = { /* FUNCTION */
+  { RGB_TOG, RGB_MOD, RGB_HUI, RGB_HUD, RGB_SAI, RGB_SAD, RGB_VAI, RGB_VAD, BL_TOGG, BL_INC,  BL_DEC,  _______, _______, _______, _______  },
+  { _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, KC_UP  , _______, KC_PSCR, _______  },
+  { _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, KC_LEFT, KC_DOWN, KC_RGHT, _______, _______  },
+  { _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______  },
+  { RESET  , _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______  },
+ },
+};
+
+const uint16_t PROGMEM fn_actions[] = {
```
Comment (noroadsleft) [inline keyboards/xd75/keymaps/martin_cennum/keymap.c]: This code block can be safely deleted.

File: keyboards/xd75/keymaps/msiu/keymap.c [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+ * |--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------|
+ * | ESC    | A      | S      | D      | F      | G      | 4      | 5      | 6      | H      | J      | K      | L      | ;      | '      |
+ * |--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------|
+ * | LSHIFT | Z      | X      | C      | V      | B      | 1      | 2      | 3      | N      | M      | ,      | .      | /      | ENTER  |
+ * |--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+-----------------+--------+--------|
+ * | LCTRL  | HYPER  | LALT   | LGUI   | LOWER  | FN     | 0      |        | .      | SPACE  | RAISE  | LEFT   | DOWN   | UP     | RIGHT  |
+ * '--------------------------------------------------------------------------------------------------------------------------------------'
+ */
+
+ [_QW] = { /* QWERTY */
```
Comment (noroadsleft) [inline keyboards/xd75/keymaps/msiu/keymap.c]: All your layers should be formatted to use a layout macro, like so:

https://github.com/qmk/qmk_firmware/blob/d9453a93f34a71f5dd5bf8cd2ef75fe21a2dcc7b/keyboards/planck/keymaps/default/keymap.c#L57-L62

File: keyboards/xd75/keymaps/msiu/keymap.c [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+ [_FN] = { /* FUNCTION */
+  { RGB_TOG, RGB_MOD, RGB_HUI, RGB_HUD, RGB_SAI, RGB_SAD, RGB_VAI, RGB_VAD, BL_TOGG, BL_INC,  BL_DEC,  _______, _______, _______, _______  },
+  { _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, KC_UP  , _______, KC_PSCR, _______  },
+  { _______, _______, _______, TO(_DV), TO(_QW), _______, _______, _______, _______, _______, KC_LEFT, KC_DOWN, KC_RGHT, _______, _______  },
+  { _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______  },
+  { RESET  , _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______  },
+ },
+};
+
+const uint16_t PROGMEM fn_actions[] = {
```
Comment (noroadsleft) [inline keyboards/xd75/keymaps/msiu/keymap.c]: Lines 123-143 can all be deleted.

The functions are deprecated, and this keymap isn't using the functionality they provide.

File: keyboards/xd75/keymaps/msiu/rules.mk [hunk 20]
```diff
@@ -0,0 +1,5 @@
+
```
Comment (noroadsleft) [inline keyboards/xd75/keymaps/msiu/rules.mk]: This file can be deleted.

The code block here is left over from a previous build script, and is no longer needed.

File: nano.save [hunk 21]
```diff
@@ -0,0 +1 @@
+',.pasdfqwer
```
Comment (noroadsleft) [inline nano.save]: This file should be removed.

**Comment by drashna** [comment]
Any updates on this?

**Comment by drashna** [comment]
Thanks!

**Pull request merged**